HIVE-24397 : Add the projection specification to the table request object and add placeholders in ObjectStore.java

vnhive opened the pull request:
HIVE-24397 : Add the projection specification to the table request object and add placeholders in ObjectStore.java

The GetProjectionsSpec object is added to GetTablesRequest. The projection specification are passed down
to the RawStore implementations to enable creating a subset of the columns that are returned.

This patch does not introduce any customer facing changes.

vihangk1 commented on standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/client/TestTablesGetExists.java:
```diff
@@ -402,6 +393,134 @@ public void testGetTableObjectsByName() throws Exception {
+    request.setDbName(DEFAULT_DATABASE);
+
+    GetProjectionsSpec projectSpec = request.getProjectionSpec();
+    List<String> projectedFields = Arrays.asList("database", "tableName", "createTime", "lastAccessTime");
```
can we add some nested single valued fields here too. Something other than sd.location? like for example. sd.serdeInfo.serializationLib? Also, what is the expected behavior if the user adds a multi-valued field like sd.parameters. Can we have some tests around that too.

vihangk1 commented on standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java:
```diff
@@ -1855,43 +1869,81 @@ private MTable getMTable(String catName, String db, String table) {
+      if (projectionSpec == null) {
+        mtables = (List<MTable>) query.execute(db, catName, lowered_tbl_names);
+      }
+      else if(projectionSpec.getFieldList() != null && projectionSpec.getFieldList().size() > 1) {
```
why do we need to code blockes for projectionSpec fieldList size > 1 and == 1? the code looks very similar.

vihangk1 commented on standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java:
```diff
@@ -1855,43 +1869,81 @@ private MTable getMTable(String catName, String db, String table) {
+      }
+      else if(projectionSpec.getFieldList() != null && projectionSpec.getFieldList().size() > 1) {
+        // fetch partially filled tables using result clause
+        query.setResult(Joiner.on(',').join(projectionSpec.getFieldList()));
```
Can we add a validation method before actually executing the query? That way you don't hit the database query if the projectionSpec has invalid fields and throw an error early. The performance is better and code becomes cleaner since you avoid the logic at 1924-1915 if you validate the fields first.

vihangk1 commented on standalone-metastore/metastore-common/src/main/thrift/hive_metastore.thrift:
```diff
@@ -1454,7 +1454,8 @@ struct GetTablesRequest {
   5: optional list<string> processorCapabilities,
-  6: optional string processorIdentifier
+  6: optional string processorIdentifier,
+  7: optional GetProjectionsSpec projectionSpec
```
can you add a comment here saying what is support and what is not? For example, ProjectionSpec has fields to filter the parameters too. But we don't support it here. The code should also validate such cases and throw a MetaException with appropriate error message.

vihangk1 commented on standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/IMetaStoreClient.java:
```diff
@@ -733,6 +733,27 @@ Table getTable(String catName, String dbName, String tableName,
+   * @throws MetaException
+   *          Any other errors
+   */
+  List<Table> getTableObjectsByRequest(GetTablesRequest request)
```
Does the method name have to include byRequest? Its self-evident by the signature right? How about getTables(GetTablesRequest request). Also, can we keep the return type as GetTablesResponse here so that the interface is more extendable in the future?

vihangk1 commented on standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java:
```diff
@@ -2360,6 +2360,20 @@ public Table getTable(String catName, String dbName, String tableName, String va
   }
 
+  @Override
+  public List<Table> getTableObjectsByRequest(GetTablesRequest req) throws TException {
```
Do you need to do any special handling of SessionMetastoreClient?

vnhive replied:
You are referring to SessionHiveMetaStoreClient right ?

SessionHiveMetaStoreClient does not have an implementation for get_partitions_with_specs and piggybacks on the implementation in its superclass (HiveMetaStoreClient) from the inheritance hierarchy. I just followed the same pattern here.

Also it just returns a list of table objects, basically a read query and should work the same across sessions, since, it just returns persisted session information.

vihangk1 reviewed (commented):
Left some suggestions below. Overall looks good to me.

vnhive commented:
In the latest pull request I have addressed the comment to change the API to getTables and also return a GetTablesResult object instead of List<Table>

vihangk1 reviewed (commented):
The patch looks good to me. I just some minor comments below. Once you address those, I can approve this PR for the merge.

vnhive commented:
@vihangk1 Addressed all your comments in this push.

nrg4878 commented on standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/IMetaStoreClient.java:
```diff
@@ -733,6 +733,27 @@ Table getTable(String catName, String dbName, String tableName,
+   * @throws MetaException
+   *          Any other errors
+   */
+  GetTablesResult getTables(GetTablesRequest request)
```
This method should have a different signature. So the Request/Result objects are serialized/deserialized between the client and the server. The client side application code (for example HiveServer2) should not have to deal with Request/Response objects.
So HS2 will invoke something like this
IMetastoreClient.getTableObjectsByName(String catName, String dbName, List<String> tableNames)
or
IMetastoreClient.getTableObjectsByName(String catName, String dbName, List<String> tableNames, GetProjectionSpec projectionSpec)

These 2 methods internally have to build a GetTablesRequest object based on the arguments, make the thrift call which returns the GetTablesResults object, which then have to be returned to calls as List<Table>

So the signature should be something like this
List<Table> getTables/getTableObjectsByName(catName, dbName, tableNames, projectionSpec)

vnhive replied:
I agree to this.

Request/Response objects are thrift serialization objects and are not meant for use outside the HiveMetaStore (client and server).

I also agree to your argument that if more fields are added to them tomorrow, these will need to be accounted for across all its usages.

I have changed this and updated the pull request.

nrg4878 commented on standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java:
```diff
@@ -2360,6 +2360,21 @@ public Table getTable(String catName, String dbName, String tableName, String va
+      req.setProcessorCapabilities(new ArrayList<String>(Arrays.asList(processorCapabilities)));
+    if (processorIdentifier != null)
+      req.setProcessorIdentifier(processorIdentifier);
+    if (req.getCapabilities() == null) {
```
duplicate setCapabilities() call.

vnhive commented:
@nrg4878 The change to the signature of this method was proposed by @vihangk1  and I agreed with him because the corresponding partition method returns a GetPartitionResponse. Please see the corresponding method below.

GetPartitionResponse getPartitionRequest(GetPartitionRequest req) throws NoSuchObjectException, MetaException, TException;

nrg4878 commented on standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java:
```diff
@@ -1855,43 +1869,89 @@ private MTable getMTable(String catName, String db, String table) {
+      }
+
+      if (mtables == null || mtables.isEmpty()) {
+        verifyDBExists(catName, db);
```
feels like we should just call getDatabaseInternal() and remove this refactored code verifyDBExists() code altogether. verifyDBExists() only fetches a single column (dbName) in a resultset of 1 row if the DB exists where as getDatabaseInternal() fetches few additional columns in a resultset of 1 row. I dont think it would be a huge performance degradation but the code will be bit cleaner. What do you think?

nrg4878 commented on standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java:
```diff
@@ -1839,13 +1839,27 @@ private MTable getMTable(String catName, String db, String table) {
+    dbExistsQuery.closeAll();
+  }
+
+  private List<Table> getTableObjectsByNameUtil(String catName, String db, List<String> tbl_names,
```
instead of adding getTableObjectsByNameUtil() method thats called by both flavors of getTableObjectsByName(), should we use
1) getTableObjectsByName(String catName, String db, List<String> tbl_names,
          GetProjectionsSpec projectionsSpec)

and then 
2) getTableObjectsByName(String catName, String db, List<String> tbl_names)
should invoke #1 with a null projectionSpec value

nrg4878 commented on standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java:
```diff
@@ -1912,19 +1972,19 @@ private Table convertToTable(MTable mtbl) throws MetaException {
-        convertToFieldSchemas(mtbl.getPartitionKeys()), parameters,
-        mtbl.getViewOriginalText(), mtbl.getViewExpandedText(), tableType);
+    final Table t = new Table(mtbl.getTableName(), mtbl.getDatabase() != null ? mtbl.getDatabase().getName() : null,
+            mtbl.getOwner(), mtbl.getCreateTime(), mtbl.getLastAccessTime(), mtbl.getRetention(),
```
nit: remove additional spacing

vnhive reviewed (commented):
Thank you for the comments. I have addressed all of them.

nrg4878 reviewed (approved):
@vnhive Change looks good to me. +1.
Could please create a sub-task for this jira to also provide a CachedStore implementation for this optimization. Metadata caching combined with projection filter support could improve performance dramatically. Thank you

nrg4878 commented:
Fix has been merged to master. Thank you @vnhive
